app/lib/hoppit.rb

```diff
@@ -0,0 +1,183 @@
+# frozen_string_literal: true
+
+module Hoppit
+  class Error < StandardError; end
+
+  class << self
+    def import_row(csv_row)
+      trainee = build_trainee(csv_row)
+      degree = build_degree(trainee, csv_row)
+
+      trainee.course_code = find_course(trainee, csv_row).code
+
+      trainee.save!
+      degree.save!
+    end
+
+    def build_degree(trainee, csv_row)
+      if csv_row["Country (Non UK) degree"].blank?
+        build_uk_degree(trainee, csv_row)
+      else
+        build_non_uk_degree(trainee, csv_row)
+      end
+    end
+
+    def build_uk_degree(trainee, row)
+      Degree.new(trainee: trainee).tap do |degree|
+        degree.locale_code = :uk
+        degree.uk_degree = validate_uk_degree(row["Degree type"])
+        degree.grade = to_degree_grade(row["Degree grade"])
+        degree.graduation_year = row["Graduation year"]
+        degree.institution = row["Institution"]
+        degree.subject = validate_degree_subject(row["Subject of UG. degree"])
+      end
+    end
+
+    def build_non_uk_degree(trainee, row)
+      Degree.new(trainee: trainee).tap do |degree|
+        degree.locale_code = :non_uk
+        degree.country = row["Country (Non UK) degree"]
+        degree.non_uk_degree = validate_enic_non_uk_degree(row["UK ENIC equivalent (Non UK)"])
+        degree.subject = validate_degree_subject(row["Subject of UG. Degree (Non UK)"])
+        degree.graduation_year = Date.parse(row["Undergrad degree date obtained (Non UK)"]).year
+      end
+    end
+
+    def find_course(trainee, csv_row)
+      potential_courses = trainee.provider.courses.where(
+        duration_in_years: csv_row["Duration"].scan(/\d+/).first.to_i,
+        start_date: Date.parse(csv_row["Course start date"]),
+        name: csv_row["ITT Subject 1"],
+        route: trainee.training_route,
+      )
+
+      case potential_courses.count
+      when 0
+        raise Error, "No course found"
+      when 1
+        potential_courses.take
+      else
+        raise Error, "Course ambiguous, multiple found"
+      end
+    end
+
+    def build_trainee(csv_row)
+      trainee = Trainee.new
+
+      # TODO: Temporary, the provider code will need to be added to this csv
+      trainee.provider = Provider.first
```

Comment:
I see what you mean now about the provider code. I think all of these trainees will be under the Teach First provider. We'll need to add that. It won't need to be in the CSV.